irctest/server_tests/redact.py

```diff
@@ -0,0 +1,672 @@
+        self.assertMessageMatch(
+            alice_redact,
+            command="REDACT",
+            params=[bob, msgid],
+            prefix=StrRe(f"{alice}!.*"),
+        )
+        self.assertMessageMatch(
+            bob_redact,
+            command="REDACT",
+            params=[bob, msgid],
+            prefix=StrRe(f"{alice}!.*"),
+        )
+
+    @cases.mark_capabilities(
+        "message-tags",
+        "server-time",
+        "echo-message",
+        "batch",
+        "labeled-response",
+        REDACT_CAP,
+    )
+    def testRedactedMessageDisappearsFromChathistory(self):
+        """Test that redacted messages are excluded from CHATHISTORY responses.
+
+        Per the spec, after a message is redacted, CHATHISTORY responses SHOULD either:
+        - Exclude it entirely, OR
+        - Include a REDACT message after the redacted message
+        """
+        alice = random_name("alice")
+        bob = random_name("bob")
+        channel = random_name("#channel")
+
+        # Connect alice
+        self.connectClient(
+            alice,
+            name=alice,
+            capabilities=[
+                "message-tags",
+                "server-time",
+                "echo-message",
+                "batch",
+                "labeled-response",
+                "draft/chathistory",
+                REDACT_CAP,
+            ],
+            skip_if_cap_nak=True,
+        )
+        self.joinChannel(alice, channel)
+        self.getMessages(alice)
+
+        # Connect bob
+        self.connectClient(
+            bob,
+            name=bob,
+            capabilities=[
+                "message-tags",
+                "server-time",
+                "echo-message",
+                "batch",
+                "labeled-response",
+                "draft/chathistory",
+                REDACT_CAP,
+            ],
+        )
+        self.joinChannel(bob, channel)
+        self.getMessages(alice)
+        self.getMessages(bob)
+
+        # Alice sends a message
+        self.sendLine(alice, f"PRIVMSG {channel} :This message will be redacted")
+        echo = self.getMessage(alice)
+        delivery = self.getMessage(bob)
+
+        self.assertMessageMatch(
+            echo, command="PRIVMSG", params=[channel, "This message will be redacted"]
+        )
+        self.assertMessageMatch(
+            delivery,
+            command="PRIVMSG",
+            params=[channel, "This message will be redacted"],
+        )
+
+        msgid = echo.tags.get("msgid")
+        assert msgid, "Server did not send a msgid tag"
+
+        # Bob retrieves chat history and should see the message
+        self.sendLine(bob, f"CHATHISTORY LATEST {channel} * 10")
+        messages = self.getMessages(bob)
+
+        # Extract PRIVMSG messages from the batch
+        privmsgs_before = [
+            msg
+            for msg in messages
+            if msg.command == "PRIVMSG" and channel in msg.params
+        ]
```

Comment:
Move `getBatchMessages` from `server_tests/metadata_2.py` to `cases.py`, give it a `batch_type` parameter, and use it here. It will be more reliable.